lib/service/connect.go

```diff
@@ -684,6 +687,114 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
 	return connector, nil
 }
 
+func (process *TeleportProcess) initOpenSSH() {
+	process.RegisterWithAuthServer(types.RoleNode, SSHIdentityEvent)
+	process.SSHd = openssh.NewSSHD()
+	process.RegisterCriticalFunc("common.rotate_openssh", process.syncOpenSSHRotationState)
+}
+
+func (process *TeleportProcess) syncOpenSSHRotationState() error {
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return nil
+	}
+	conn, err := process.WaitForConnector(SSHIdentityEvent, nil)
+	if conn == nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+
+	status, err := process.syncRotationStateAndBroadcast(conn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	id, err := process.storage.ReadIdentity(auth.IdentityCurrent, types.RoleInstance)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cas, err := conn.Client.GetCertAuthorities(ctx, types.OpenSSHCA, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	keysDir := process.Config.OpenSSH.KeysDir
+	if err := openssh.TryCreateDefaultAgentlesKeysDir(keysDir); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := openssh.WriteKeys(keysDir, id, cas); err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = process.SSHd.UpdateConfig(openssh.SSHdConfigUpdate{
+		SSHdConfigPath:         process.Config.OpenSSH.SSHDConfigPath,
+		TeleportSSHDConfigPath: openssh.AgentlessSSHDConfigPath,
+		OpenSSHCAPath:          filepath.Join(keysDir, openssh.TeleportOpenSSHCA),
+		HostKeyPaths:           []string{filepath.Join(keysDir, openssh.TeleportKey)},
+		HostCertPaths:          []string{filepath.Join(keysDir, openssh.TeleportCert)},
+	}, process.Config.OpenSSH.RestartSSHD)
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	state, err := process.storage.GetState(types.RoleNode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	mostRecentRotation := state.Spec.Rotation
+	if !status.needsReload {
+		for _, ca := range cas {
+			if ca.GetRotation().LastRotated.After(mostRecentRotation.LastRotated) {
+				mostRecentRotation = ca.GetRotation()
+			}
+		}
+	}
+
+	if err := process.storage.WriteState(types.RoleNode, *state); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := registerServer(process.Config, ctx, conn.Client, mostRecentRotation); err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		process.Shutdown(ctx)
+		if err := process.Close(); err != nil {
```

Comment:
I don't think we need a `process.Close` after `process.Shutdown`, do we?

Reply:
Do we understand the difference between these two functions? Why is one used some places and the other in others? Can/should they be consolidated?

Reply:
I wonder if it's some awkward thing caused by using the `ExitContext` as the shutdown context here; what happens if you pass `context.Background()` to `process.Shutdown` instead?

Reply:
Isn't the `GracefulExitContext` cancelled at the very beginning of `Shutdown`?

Reply:
Yeah, this is equivalent of passing in a context to `Shutdown` and immediately closing it, as `(*LocalSupervisor).BroadcastEvent` will close the graceful exit context as soon as it receives the `TeleportExitEvent`.

What do the logs show if you just pass `context.Background()` to `process.Shutdown()`, which is what you'd effectively get with an external signal? The supervisor should periodically show a list of services that are waiting to exit.

Reply:
The easiest solution to what I think is actually an oversight in the interaction between `WaitAndReload` and shutdown might be to do
```go
ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
// (or just context.WithCancel, but having a timeout is actually more correct here)
defer cancel()
process.Shutdown(ctx)
```
as canceling the context passed to `Shutdown` should also result in closing the `process.exitContext`, which should unblock `WaitForSignal`.